src/EditorFeatures/Core/GoToDefinition/GoToDefinitionCommandHandler.cs

```diff
@@ -2,88 +2,84 @@
             }
 
-            nextHandler();
+            return false;
         }
 
         // Internal for testing purposes only.
-        internal bool TryExecuteCommand(ITextSnapshot snapshot, int caretPosition)
-            => TryExecuteCommand(snapshot.GetOpenDocumentInCurrentContextWithChanges(), caretPosition);
+        internal bool TryExecuteCommand(ITextSnapshot snapshot, int caretPosition, CommandExecutionContext context)
+            => TryExecuteCommand(snapshot.GetOpenDocumentInCurrentContextWithChanges(), caretPosition, context);
 
-        internal bool TryExecuteCommand(Document document, int caretPosition)
-            => TryExecuteCommand(document, caretPosition, document.GetLanguageService<IGoToDefinitionService>());
+        internal bool TryExecuteCommand(Document document, int caretPosition, CommandExecutionContext context)
+            => TryExecuteCommand(document, caretPosition, document.GetLanguageService<IGoToDefinitionService>(), context);
 
-        internal bool TryExecuteCommand(Document document, int caretPosition, IGoToDefinitionService goToDefinitionService)
+        internal bool TryExecuteCommand(Document document, int caretPosition, IGoToDefinitionService goToDefinitionService, CommandExecutionContext context)
         {
             string errorMessage = null;
 
-            var result = _waitIndicator.Wait(
-                title: EditorFeaturesResources.Go_to_Definition,
-                message: EditorFeaturesResources.Navigating_to_definition,
-                allowCancel: true,
-                action: waitContext =>
+            using (context.WaitContext.AddScope(allowCancellation: true, EditorFeaturesResources.Navigating_to_definition))
+            {
+                if (goToDefinitionService != null &&
+                    goToDefinitionService.TryGoToDefinition(document, caretPosition, context.WaitContext.UserCancellationToken))
                 {
-                    if (goToDefinitionService != null &&
-                        goToDefinitionService.TryGoToDefinition(document, caretPosition, waitContext.CancellationToken))
-                    {
-                        return;
-                    }
+                    return true;
+                }
 
-                    errorMessage = EditorFeaturesResources.Cannot_navigate_to_the_symbol_under_the_caret;
-                });
+                errorMessage = EditorFeaturesResources.Cannot_navigate_to_the_symbol_under_the_caret;
+            }
 
-            if (result == WaitIndicatorResult.Completed && errorMessage != null)
+            if (errorMessage != null)
             {
+                // We are about to show a modal UI dialog so we should take over the command execution
+                // wait context. That means the command system won't attempt to show its own wait dialog 
+                // and also will take it into consideration when measuring command handling duration.
+                context.WaitContext.TakeOwnership();
```

Comment:
>context.WaitContext.TakeOwnership(); [](start = 16, length = 36)

it's not strictly necessary as threaded wait dialog knows to hide itself when any other dialog is shown, but it's not perfect and there is still a chance for 2 dialogs to be shown for some time interval. So whenever appropriate command handlers should call this before showing modal dialogs.